### ISSUE TITLE
Using `.authenticate` with a slow hasher can stall the event loop

### ISSUE THREAD

**Issue by MrMage** [vapor/auth#38]
I am using the following call to authenticate the user:

    User.authenticate(
        username: username,
        password: password,
        using: passwordVerifier,
        on: $0)

This runs `guard try verifier.verify(basic.password, created: user.basicPassword) else {` after fetching the user's encrypted password. If the verifier is slow, e.g. `bcrypt`-based, this verification might take up to a few 100 milliseconds (but even 50-100 would already be bad). During that time, the event loop can not process any other events, meaning that any other operations on the same loop (e.g. other requests, DB calls, etc.) get delayed by this much. In particular, authenticating one user's request might noticeably delay another user's request just by the virtue of being scheduled on the same event loop.

To avoid this, I'd suggest running `verifier.verify` off the event loop (possibly on a concurrent dispatch queue) and, once that is done, fulfilling a subsequent promise back on the event loop.

What do you think? Also, which way of moving this work off the event loop would you prefer? In fact, should Vapor offer a "standard"/"best practice" method of doing this?

**Comment by tanner0101** [comment]
This one has been brought up before. I think I'm in favor of doing the hashing on a background thread. It's still CPU-bound work at the end of the day, but at least if it's on a background thread the scheduler can make sure other tasks aren't getting starved. 

We should take advantage of NIO's `BlockingIOThreadPool` here to ensure only a limited number of threads are spun up. From there, as long as we create a promise with the current event loop, completing that promise is guaranteed to return us to the event loop once the hash is done.

**Comment by MrMage** [comment]
Thanks for the feedback. How should the `BlockingIOThreadPool` be provided? Should there be a service for it that we could use here?

**Comment by tanner0101** [comment]
Yup, that is actually being added in https://github.com/vapor/vapor/pull/1647. Will be included in the 3.0.0 tag. `req.make(BlockingIOThreadPool.self)` should do the trick.

**Comment by MrMage** [comment]
Thanks! I'll look into using that once it is merged.

**Comment by tanner0101** [comment]
1: Maybe we can make this part of the PasswordVerifier instead. If we open up that protocol to returning a Future, then we can require a `BlockingIOThreadPool` while initializing the individual password verifier (maybe even create a generic "nonblocking password verifier" wrapper).

2: Yup, that's bad. Nice catch. Will fix that in https://github.com/vapor/vapor/pull/1663.

3: I don't see a good way around this. In terms of `DatabaseConnectable` usage with `Request`, the connection is reserved for the entire lifetime of the request. This is necessary since developers will assume that any state-affecting queries they perform will make sense in the context of a single request. In terms of passing in a concrete connection as the `DatabaseConnectable`, it's possible that this protocol may need to do more DB work _after_ the hash which would be impossible if we created some sort of early-releasing connection API here. I see the reasoning here but think this is a tangential concern and something that will take a good amount of thought.

**Comment by 0xTim** [comment]
Closing due to inactivity - feel free to reopen!

**Comment by MrMage** [comment]
I think this is still relevant and would affect any Vapor project servicing a high number of logins (to the tune of a few per second).
